device/transport/amqp/src/amqp_twin_client.ts

```diff
@@ -139,40 +141,40 @@ export class AmqpTwinClient extends EventEmitter {
-                    this._endpoint = endpoint.moduleTwinPath(credentials.deviceId, credentials.moduleId);
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_06_007: [The endpoint argument for attachReceiverLink shall be `/device/<deviceId>/twin`.] */
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_18_001: [If a `moduleId` value was set in the device's connection string, the endpoint argument for `attachReceiverLink` shall be `/devices/<deviceId>/modules/<moduleId>/twin`]*/
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_06_009: [The endpoint argument for attachSenderLink shall be `/device/<deviceId>/twin`.] */
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_18_002: [If a `moduleId` value was set in the device's connection string, the endpoint argument for `attachSenderLink` shall be `/device/<deviceId>/modules/<moduleId>/twin`.]*/
+                if (credentials.moduleId) {
+                  this._endpoint = endpoint.moduleTwinPath(credentials.deviceId, credentials.moduleId);
+                } else {
+                  this._endpoint = endpoint.deviceTwinPath(credentials.deviceId);
+                }
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_06_006: [When a listener is added for the `response` event, and the `post` event is NOT already subscribed, upstream and downstream links are established via calls to `attachReceiverLink` and `attachSenderLink`.] */
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_06_012: [When a listener is added for the `post` event, and the `response` event is NOT already subscribed, upstream and downstream links are established via calls to `attachReceiverLink` and `attachSenderLine`.] */
+                /*Codes_SRS_NODE_DEVICE_AMQP_TWIN_16_036: [The same correlationId shall be used for both the sender and receiver links.]*/
+                const linkCorrelationId: string  = uuid.v4().toString();
+                this._client.attachSenderLink( this._endpoint, this._generateTwinLinkProperties(linkCorrelationId), (senderLinkError?: Error, senderTransportObject?: any): void => {
+                  if (senderLinkError) {
+                      /* Codes_SRS_NODE_DEVICE_AMQP_TWIN_06_022: [If an error occurs on establishing the upstream or downstream link then the `error` event shall be emitted.] */
+                      this._fsm.transition('detached', senderLinkError, attachCallback);
                   } else {
-                    this._endpoint = endpoint.deviceTwinPath(credentials.deviceId);
+                    this._senderLink = senderTransportObject;
+                    this._senderLink.on('error', this._errorHandler);
+                    this._client.attachReceiverLink( this._endpoint, this._generateTwinLinkProperties(linkCorrelationId), (receiverLinkError?: Error, receiverTransportObject?: any): void => {
+                      if (receiverLinkError) {
+                        this._fsm.transition('detached', receiverLinkError, attachCallback);
```

Comment:
>this._fsm.transition('detached', receiverLinkError, attachCallback); [](start = 24, length = 68)

This probably predates your code, but if the receiver link fails to attach, it will leave the sender link in an attached state.  Does this cause any problems?